src/array/cuda/rowwise_sampling.cu

```diff
@@ -232,27 +233,372 @@ __global__ void _CSRRowWiseSampleReplaceKernel(
+            const int num = curand(&rng) % num_picks;
+            // use max so as to achieve the replacement order the serial
+            // algorithm would have
+            AtomicMax(out_idxs+out_row_start+num, idx);
+          }
+        }
+        __syncthreads();
+
+        weights_sum += block_aggregate;
+      }
+
+      // copy permutation over
+      for (int idx = threadIdx.x; idx < num_picks; idx += BLOCK_SIZE) {
+        const IdType perm_idx = out_idxs[out_row_start+idx]+in_row_start;
+        out_rows[out_row_start+idx] = row;
+        out_cols[out_row_start+idx] = in_index[perm_idx];
+        out_idxs[out_row_start+idx] = data ? data[perm_idx] : perm_idx;
+      }
+    }
+
+    out_row += 1;
+  }
+}
+
+/**
+* @brief Perform row-wise sampling on a CSR matrix, and generate a COO matrix,
+* with replacement.
+*
+* @tparam IdType The ID type used for matrices.
+* @tparam FloatType The float type used for probability.
+* @tparam TILE_SIZE The number of rows covered by each threadblock.
+* @tparam MAX_FANOUT Max fanout allowed.
+* @param rand_seed The random seed to use.
+* @param num_picks The number of non-zeros to pick per row.
+* @param num_rows The number of rows to pick.
+* @param in_rows The set of rows to pick.
+* @param in_ptr The indptr array of the input CSR.
+* @param in_index The indices array of the input CSR.
+* @param data The data array of the input CSR.
+* @param out_ptr The offset to write each row to in the output COO.
+* @param out_rows The rows of the output COO (output).
+* @param out_cols The columns of the output COO (output).
+* @param out_idxs The data array of the output COO (output).
+*/
+template<typename IdType, typename FloatType, int TILE_SIZE, int MAX_FANOUT>
+__global__ void _CSRRowWiseSampleReplaceKernel(
+    const uint64_t rand_seed,
+    const int64_t num_picks,
+    const int64_t num_rows,
+    const IdType * const in_rows,
+    const FloatType * const prob,
+    const IdType * const in_ptr,
+    const IdType * const in_index,
+    const IdType * const data,
+    const IdType * const out_ptr,
+    IdType * const out_rows,
+    IdType * const out_cols,
+    IdType * const out_idxs) {
+  // we assign one warp per row
+  assert(blockDim.x == BLOCK_SIZE);
+
+  int64_t out_row = blockIdx.x*TILE_SIZE+threadIdx.y;
+  const int64_t last_row = min(static_cast<int64_t>(blockIdx.x+1)*TILE_SIZE, num_rows);
+
+  curandStatePhilox4_32_10_t rng;
+  curand_init((rand_seed*gridDim.x+blockIdx.x)*blockDim.y+threadIdx.y, threadIdx.x, 0, &rng);
+
+  while (out_row < last_row) {
+    const int64_t row = in_rows[out_row];
+
+    const int64_t in_row_start = in_ptr[row];
+    const int64_t out_row_start = out_ptr[out_row];
+
+    const int64_t deg = in_ptr[row+1] - in_row_start;
+
+    if (deg > 0) {
+      /*
+      * The algorithm here is a simple sorted inverse transform sampling.
+      * To parallelize it, we use moving windows for the random numbers
+      * and the CDF (prefix sum here).
+      */
```

Comment:
We can put the description of the design choices at the beginning in the @brief section.

Reply:
Changed accordingly.